### PR TITLE
[Fault Injection] Add pre-training failure taxonomy and validation

### PR DESCRIPTION
**Pull request by zhuangwang93**
## Summary

- add the public SystemFailureType taxonomy for 21 canonical pre-training failure causes
- validate cause/effect compatibility while preserving backward compatibility when the taxonomy is omitted
- propagate the taxonomy through fault records, serialization, reports, public exports, tests, and documentation
- document the two-stage A100 fault-injection campaign and clearly separate exact SCOUT localization from executor/sandbox ground truth
- fix the torchrun rendezvous entry point so torchrun can instantiate it as a zero-argument factory

## Hardware validation

- Stage 1, two A100 hosts: 48/48 scheduled occurrences and 53/53 injected actions were detected and localized across the six implemented injectable effects
- Stage 2, one A100 host with four active and four standby GPUs: all 21 canonical failure causes were exercised and recovered (17 restarts, 4 replacements)
- Stage 2 exact SCOUT localization was independently established for tensor corruption; the other 20 cases use executor/sandbox ground truth and are documented as such

## Tests

- 379 passed: tests/unit/test_fault_injection.py and tests/unit/test_unified_api.py
- 48 passed: focused torchrun and compatibility tests
- strict MkDocs build passed
- both documented A100 campaigns completed with no remaining GPU processes

### CODE REVIEW

File: lm_resiliency/fault_injection/reports.py [hunk 1]
```diff
@@ -43,6 +43,7 @@ class FaultInjectionRecord:
     execution_rank: int
     target: Mapping[str, Any]
     parameters: Mapping[str, Any]
+    system_failure_type: str | None = None
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply lm_resiliency/fault_injection/reports.py]: Fixed in 6e10fe4. system_failure_type now follows every pre-existing public field, preserving the legacy positional constructor. Added a regression test that supplies all pre-existing fields positionally and verifies their original assignments.

File: lm_resiliency/fault_injection/reports.py [hunk 2]
```diff
@@ -135,6 +137,9 @@ def to_dict(self) -> dict[str, Any]:
                 "evidence": thaw_json(self.evidence),
                 "error": self.error,
             }
+            if self.system_failure_type is not None:
+                value["system_failure_type"] = self.system_failure_type
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply lm_resiliency/fault_injection/reports.py:141]: Fixed in 6e10fe4. The artifact comparator now requires exact system_failure_type field presence and value agreement with the authenticated manifest. Regression coverage rejects both a missing record value and a mismatched value.

File: lm_resiliency/fault_injection/config.py [hunk 3]
```diff
@@ -40,7 +41,14 @@
     "path",
     "metadata",
 }
-_FAULT_KEYS = {"fault_id", "id", "type", "target", "parameters"}
+_FAULT_KEYS = {
+    "fault_id",
+    "id",
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply lm_resiliency/fault_injection/config.py:48]: Fixed in 818fb1f. SCHEMA_VERSION is now 2, current writers and generated manifests emit version 2, and readers continue accepting explicit or omitted legacy version-1 manifests. Version 1 is rejected if system_failure_type is present. Added regression tests for current, explicit legacy, omitted legacy, and invalid version-1 root-cause manifests; updated docs and checked-in generated campaigns.

File: lm_resiliency/fault_injection/config.py [hunk 4]
```diff
@@ -12,9 +12,12 @@
 from pathlib import Path
 from typing import Any, Mapping
 
+from lm_resiliency.failure_types import SystemFailureType
 from lm_resiliency.fault_injection._json import freeze_json_mapping, thaw_json
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply lm_resiliency/fault_injection/config.py:16]: Fixed in 0e1bdcd. FaultCampaign now defaults unchanged programmatic campaigns to schema 1, preserving their serialized manifest identity and existing CampaignJournal bindings. Schema 2 remains the latest supported version and is explicitly required when system_failure_type is used. Added a journal-binding regression test.

File: examples/torchrun/resiliency_cycle/campaigns/single_node_pressure.json [hunk 5]
```diff
@@ -653,6 +653,6 @@
     "total_steps": 42
   },
   "name": "torchrun-resiliency-cycle-all-failure-types",
-  "schema_version": 1,
+  "schema_version": 2,
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply examples/torchrun/resiliency_cycle/campaigns/single_node_pressure.json]: Fixed in 0e1bdcd. Restored single_node_pressure.json (and the other unchanged generated campaigns) to schema 1. Its SHA-256 is again 36781f2b... and its manifest identity is again 4c9177d8..., matching the recorded Stage 2 evidence. Added exact digest and manifest-identity assertions.

File: lm_resiliency/fault_injection/config.py [hunk 6]
```diff
@@ -631,15 +831,23 @@ def from_dict(cls, value: Mapping[str, Any]) -> "FaultSpec":
             type=FailureType(value["type"]),
             target=FaultTarget.from_dict(value["target"]),
             parameters=value.get("parameters", {}),
+            system_failure_type=(
+                None
+                if value.get("system_failure_type") is None
+                else SystemFailureType(value["system_failure_type"])
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply lm_resiliency/fault_injection/config.py:837]: Fixed in da18228. FaultCampaign.from_dict now checks raw schema-1 fault mappings for system_failure_type key presence before normalization, so null and non-null values are both rejected. Added a regression test for the null case. The same batch also allowlists the two documented campaign hashes for the secret scanner.

File: tests/validation/fault_injection/compare.py [hunk 7]
```diff
@@ -717,6 +717,25 @@ def _validate_record_against_manifest(
     expected_kind = _expected_action_kind(action)
     if _required_string(record.get("expected_kind"), "injection expected_kind") != expected_kind:
         raise ValueError("injection record expected_kind does not match the authenticated manifest")
+    if ("system_failure_type" in record) != ("system_failure_type" in action):
+        raise ValueError(
+            "injection record system_failure_type does not match the authenticated manifest"
+        )
+    if "system_failure_type" in action:
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply tests/validation/fault_injection/compare.py:757]: Fixed in 0c5c899. The comparator now validates the authenticated embedded manifest schema before evaluating records: omitted schema is treated as v1, unsupported versions are rejected, and any system_failure_type field under v1 is rejected even when the record matches. The existing tampering test now uses schema v2, and regression coverage exercises both implicit and explicit schema v1.

File: lm_resiliency/fault_injection/config.py [hunk 8]
```diff
@@ -887,10 +1095,19 @@ def __post_init__(self) -> None:
...
+            supported = ", ".join(str(version) for version in sorted(_SUPPORTED_SCHEMA_VERSIONS))
             raise ValueError(
                 f"unsupported campaign schema_version {self.schema_version}; "
-                f"expected {SCHEMA_VERSION}"
+                f"expected one of: {supported}"
+            )
+        if self.schema_version == _LEGACY_SCHEMA_VERSION and any(
+            fault.system_failure_type is not None
+            for incident in self.incidents
+            for fault in incident.faults
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply lm_resiliency/fault_injection/config.py:1109]: Fixed in 0c5c899. FaultCampaign now validates normalized incident objects immediately, before schema-specific fault traversal, so invalid objects consistently raise the documented TypeError. Added a focused regression test for the schema-v1 construction path.

**Comment by zhuangwang93** [comment]
@codex review

File: tests/validation/fault_injection/compare.py [hunk 9]
```diff
@@ -512,6 +514,32 @@ def _manifest_actions(
...
+    if schema_version not in {1, SCHEMA_VERSION}:
+        raise ValueError(
+            f"unsupported injection manifest schema_version {schema_version}; "
+            f"expected one of: 1, {SCHEMA_VERSION}"
+        )
+    if schema_version == 1 and any(
+        "system_failure_type" in action
+        for actions in manifest_actions.values()
+        for action in actions.values()
+    ):
```
[the remark this answers is not in the document]
Comment (zhuangwang93) [reply tests/validation/fault_injection/compare.py:538]: Fixed in 2333002. For schema v2, every authenticated manifest action is now parsed through FaultSpec.from_dict(), which applies the canonical SystemFailureType enum and the shared cause/effect compatibility mapping before records can be evaluated. Added regressions for an unknown root cause and for cuda_out_of_memory paired with network_partition; the focused suite passes 433 tests.

**Comment by zhuangwang93** [comment]
@codex review
